firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt

```diff
@@ -0,0 +1,13 @@
+SCENARIO("Value is in provided range")
+  GIVEN("A value '\x83'")
```

Comment:
These tests should have coverage of all possible relationships between inputs and outputs:
- value < floor
- value == floor
- floor < value < ceiling
- value == ceiling
- value > ceiling

We should also have tests specifying what behavior we should expect when floor == ceiling and when floor > ceiling.

Reply:
@raavilagoo This applies not just to the `clamp` function, but also to the `within` function.

Reply:
Here's what I can see about coverage on the `clamp` function:

Test cases needed (bolded items are covered)
1. **value < floor**
2. **value == floor**
3. **floor < value < ceiling**
4. value == ceiling
5. **value > ceiling** (this is covered twice)
6. floor == ceiling
  a. value < floor
  b. **floor == ceiling, value == floor == ceiling**
  c. floor == ceiling, value > ceiling
7. floor > ceiling
  a. value < ceiling
  b. floor > ceiling, value == floor
  c. floor > ceiling, value == ceiling
  d. **floor > ceiling, value > floor**

| test case | floor | test value | ceil |
| -- | -- | -- | -- |
| 3 | 1 | 131 | 255 |
| 5 | 1 | 131 | 10 |
| 1 | 10 | 8 | 131 |
| 2 | 8 | 8 | 131 |
| 5 | 8 | 83 | 15 |
| 6b | 8 | 8 | 8 |
| 7d | 15 | 83 | 8 |

Reply:
I would suggest grouping cases 1-5 in one scenario, then grouping cases 6a-c in another scenario, then grouping cases 7a-d in another scenario.

Reply:
We also need 7e, floor > value > ceiling